Initial commit (module draft)

anlvk opened the pull request.

Commit:
Initial commit (module draft)

Commit:
Cleaning

skyWalkerBlr commented on cache_content.info.yml:
```diff
@@ -0,0 +1,9 @@
+description: Custom cache for content
+package: Custom
+version: 1.0
+core: 8.x
```
https://www.drupal.org/docs/develop/creating-modules/let-drupal-know-about-your-module-with-an-infoyml-file

skyWalkerBlr commented on src/Controller/CacheContentController.php:
```diff
@@ -0,0 +1,32 @@
+   * @return void
+   */
+  public function first_block(): array {
+    $output = [
+      '#title' => 'First Block title',
+      '#markup' => 'First Block',
+    ];
+
+    return $output;
```
an extra variable, you can return an array

skyWalkerBlr commented on src/EvenMinuteChecker.php:
```diff
@@ -0,0 +1,32 @@
+ */
+class EvenMinuteChecker {
+
+    private $currentMinute;
```
Property must have a comment

skyWalkerBlr commented on src/EvenMinuteChecker.php:
```diff
@@ -0,0 +1,32 @@
+     * Undocumented function
+     */
+    public function __construct() {
+        $this->currentMinute = date('i');
```
Please use here:
- \Drupal\Component\Datetime\TimeInterface
- \Drupal\Core\Datetime\DateFormatterInterface

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+   * @var \Drupal\cache_content\EvenMinuteChecker
+   */
+  protected $minuteChecker;
+
```
Extra line

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+
+namespace Drupal\cache_content\Plugin\Block;
+
+
```
extra line

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\Core\Form\FormStateInterface;
+
+
```
extra line

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+   *   A render array.
+   */
+  public function build():array {
+
```
extra line

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+   * @return array
+   *   A render array.
+   */
+  public function build():array {
```
space is missing for return type

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
```
return type is missing

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+    $config = $this->getConfiguration();
+
+    // Display content based on a minute when a block is loaded.
+    if ($isEvenMinute === true) {
+      $message = $config['even_text'];
+    } else {
+      $message = $config['odd_text'];
+    }
```
You can use here ternary operator

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+      '#markup' => $message,
+    ];
+  
+    return $block;
```
YOu can return array instead of creating new variable

skyWalkerBlr commented on src/Plugin/Block/CacheContentFirstBlock.php:
```diff
@@ -0,0 +1,110 @@
+  /**
+   * Undocumented function
+   *
+   * @param [type] $form
```
[type] == array?

Commit:
Fixes after review
1. Remove extra lines
2. Refactor code to implement \Drupal\Component\Datetime\TimeInterface and \Drupal\Core\Datetime\DateFormatterInterface
3. Remove extra variables
4. Change .info.yml file (correct core_version_requirement)
5. Define return types and other type hinting
6. Add/modify comments

Commit:
Add extra line.

Commit:
Major modifications
1. Remove Controller directory and class.
2. Modify documentation for each class and method: add proper documentation.
3. Configuration form is created.
4. A new block is added to display data from configuration form (CacheContextSecondBlock).
5. Custom cache context is created and implemented in CacheContextFisrtBlock.
6. A route is created to provide module configuration page.
7. Other modifications to implement new functionality above.

Commit:
Minor changes: add lines in files.

Commit:
Again minor changes: add lines in files.

Commit:
Module update:
1. Third block (CacheContentThirdBlock) is added to display nodes depending on odd/even minute.
2. Minor modifications for the first block (CacheContentFirstBlock): rename a variable according to PHP standards.

skyWalkerBlr commented on src/Cache/Context/OddEvenMinuteContext.php:
```diff
@@ -0,0 +1,74 @@
+     * @param EvenMinuteChecker $minuteChecker
+     */
+	public function __construct(EvenMinuteChecker $minuteChecker) {
+		$this->minuteChecker = $minuteChecker;
```
Code style

skyWalkerBlr replied:
Could be replaced with constructor property promotion

skyWalkerBlr commented on src/Cache/Context/OddEvenMinuteContext.php:
```diff
@@ -0,0 +1,74 @@
+     * @return string
+     */
+	public function getContext(): string {
+        $isEvenMinute = $this->minuteChecker->isCurrentMinuteEven();
+        $result = ($isEvenMinute === true) ? 'even' : 'odd';
+
+        return $result;
+    }
```
Code style

skyWalkerBlr commented on src/Cache/Context/OddEvenMinuteContext.php:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    /**
+     * Returns the label of the cache context.
+     *
+     * @return string
+     */
+	public static function getLabel(): string {
+		return t('Odd/even minute context');
+	}
+
```
code style

skyWalkerBlr commented on src/EvenMinuteChecker.php:
```diff
@@ -0,0 +1,58 @@
+     * @param TimeInterface $timeService
+     * @param DateFormatterInterface $dateFormatter
+     */
+    public function __construct(TimeInterface $timeService, DateFormatterInterface $dateFormatter) {
+        $this->timeService = $timeService;
+        $this->dateFormatter = $dateFormatter;
+    }
```
Can be replaced with property promotion constructor

skyWalkerBlr commented on src/Plugin/Block/CacheContentThirdBlock.php:
```diff
@@ -0,0 +1,96 @@
+   *
+   * @return array
+   */
+  public function getCacheContexts(): array {
+    return Cache::mergeContexts(
+        parent::getCacheContexts(),
+        ['odd_even_minute']
+    );
+  }
+
+}
```
Also cache tags method should be added as yo displaying here nodes

skyWalkerBlr commented on src/Plugin/Block/CacheContentThirdBlock.php:
```diff
@@ -0,0 +1,96 @@
+    $node = ($is_even_minute === true) ? $nodes[3] : $nodes[4];
+    $message = $view_builder->view($node, $view_mode);
+
+    return $message;
```
no need to create variable
return expression

skyWalkerBlr commented on src/Plugin/Block/CacheContentThirdBlock.php:
```diff
@@ -0,0 +1,96 @@
+    // Retrieve nodes from entity storage.
+    $view_mode = 'teaser';
+    $view_builder = $this->entityTypeManager->getViewBuilder('node');
+    $storage = $this->entityTypeManager->getStorage('node');
```
When you know that you will work only with nodes here, you can load needed storage in  create method

Commit:
Modifications after a review
1. Fix indentation in module files.
2. Implement constructor property promotion for some classes.
3. Add cache tags for third block (CacheContentThirdBlock).
4. Modify create() method to add node storage there for further usage.
5. Return expressions without create a variable.

Commit:
Improve style coding.

Commit:
Rendering modifications
1. A custom twig template is created for each block (3 twig templates).
2. A new file is created for the module to implement twig templates (cache_content.module).
3. Blocks' classes were modified to build() blocks using twig templates.

Commit:
Add end lines.

Commit:
Code style fixed.

Commit:
Modify module files according to Drupal standards + modify comments.

skyWalkerBlr commented on cache_content.routing.yml:
```diff
@@ -0,0 +1,7 @@
+    _form: '\Drupal\cache_content\Form\CustomTextConfigForm'
+    _title: 'Cache content settings.'
+  requirements:
+    _permission: 'administer site configuration'
```
вместо роли попробуй тут использовать кастомный каллбэк что бы проверить доступ. тут всё правильно по что бы потренироваться


https://www.drupal.org/docs/8/api/routing-system/access-checking-on-routes/advanced-route-access-checking

skyWalkerBlr commented on src/EvenMinuteChecker.php:
```diff
@@ -0,0 +1,44 @@
+    $currentTime = $this->timeService->getCurrentTime();
+    // Retrieve custom format: minute - from current timestamp.
+    $minute = $this->dateFormatter->format($currentTime, 'custom', 'i');
+    // True - even, false - odd.
```
Перед return всегда должна быть пустая строка. И коммент совсем лишний.

Перед return не нужно пустой строки только в том случае если у нас метод из одного ретурна состоит

Commit:
Add custom access check for module config form.

Commit:
Add cache context ID in a comment for easy search.
